demos/device-shadow-for-aws-iot-embedded-sdk/shadow_demo_main.c

```diff
@@ -0,0 +1,679 @@
+ * incoming message.
+ *
+ * @param[in] pxMqttContext MQTT context pointer.
+ * @param[in] pxPacketInfo Packet Info pointer for the incoming packet.
+ * @param[in] pxDeserializedInfo Deserialized information from the incoming packet.
+ */
+static void prvEventCallback( MQTTContext_t * pxMqttContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              MQTTDeserializedInfo_t * pxDeserializedInfo );
+
+/**
+ * @brief Process payload from /update/delta topic.
+ *
+ * This handler examines the version number and the powerOn state. If powerOn
+ * state has changed, it sets a flag for the main function to take further actions.
+ *
+ * @param[in] pPublishInfo Deserialized publish info pointer for the incoming
+ * packet.
+ */
+static void prvupdateDeltaHandler( MQTTPublishInfo_t * pxPublishInfo );
+
+/**
+ * @brief Process payload from /update/accepted topic.
+ *
+ * This handler examines the accepted message that carries the same clientToken
+ * as sent before.
+ *
+ * @param[in] pxPublishInfo Deserialized publish info pointer for the incoming
+ * packet.
+ */
+static void prvupdateAcceptedHandler( MQTTPublishInfo_t * pxPublishInfo );
+
+/*-----------------------------------------------------------*/
+
+static void prvupdateDeltaHandler( MQTTPublishInfo_t * pxPublishInfo )
+{
+    static uint32_t ulCurrentVersion = 0; /* Remember the latestVersion # we've ever received */
+    uint32_t ulVersion = 0U;
+    uint32_t ulNewState = 0U;
+    char * pcOutValue = NULL;
+    uint32_t ulOutValueLength = 0U;
+    JSONStatus_t result = JSONSuccess;
+
+    assert( pxPublishInfo != NULL );
+    assert( pxPublishInfo->pPayload != NULL );
+
+    LogInfo( ( "/update/delta json payload:%s.\n\n", ( const char * ) pxPublishInfo->pPayload ) );
+
+    /* The payload will look similar to this:
+     * {
+     *      "version": 12,
+     *      "timestamp": 1595437367,
+     *      "state": {
+     *          "powerOn": 1
+     *      },
+     *      "metadata": {
+     *          "powerOn": {
+     *          "timestamp": 1595437367
+     *          }
+     *      },
+     *      "clientToken": "388062"
+     *  }
+     */
+
+    /* Make sure the payload is a valid json document. */
+    result = JSON_Validate( pxPublishInfo->pPayload,
+                            pxPublishInfo->payloadLength );
+
+    if( result == JSONSuccess )
+    {
+        /* Then we start to get the version value by JSON keyword "version". */
+        result = JSON_Search( ( char * ) pxPublishInfo->pPayload,
+                              pxPublishInfo->payloadLength,
+                              "version",
+                              sizeof( "version" ) - 1,
+                              '.',
+                              &pcOutValue,
+                              ( size_t * ) &ulOutValueLength );
+    }
+    else
+    {
+        LogError( ( "The json document is invalid!!\n\n" ) );
```

Comment:
If the JSON document is invalid, would that make the demo unsuccessful?

Reply:
No, we just print out the log to make users aware of this condition.